### DISCUSSION TITLE
Backtick highlight

### DISCUSSION THREAD

**Discussion by taste5** [cagix/pandoc-thesis#56] [Q&A] [answered]
Hallo,
Ich habe dein Template auf der Suche nach einem geeigneten Format für meine Thesis gefunden. Und freue mich sehr mit deiner Vorlage gut zurecht zu kommen. Lustigerweise hat mein Prozess auch genau an der uni Bielefeld angefangen, nämlich mit Luhmann und seinem Zettelkasten und nun scheint die Uni Bielefeld mein Begleiter richtung Abgabe zu sein.
Vielen Dank.

Es gibt ein kleines Thema: Ich würde gerne highlighted Backlinks einfügen. habe es schon mit:
```
OPTIONS                += --highlight-style=pygments
OPTIONS                 = -f markdown+implicit_figures+hard_line_breaks+backtick_code_blocks
```
versucht. Leider Erfolglos. Weiß jemand woran das liegt?

**Comment by cagix** [comment]
@moerk0 Bevor ich da tiefer einsteige, drei kurze Fragen:

- Welche Variante nutzt Du? (simple vs. Eisvogel vs. Thesis?)
- Du möchtest Inline-Code (Code in einfachen Backticks) hervorgehoben haben, so wie es bei den Code-Blöcken auch gemacht wird? 
- Soll das dann einfach nur eine einfach Hervorhebung sein (beispielsweise hellgelber Hintergrund) oder soll das auch richtiges Syntaxhighlighting sein?

**Reply by taste5** [reply]
@cagix Vielen Dank für die Bereitschaft dich der Sache anzunehmnen.

Ich nutze `simple` Thesi.
Eisvogel und Thesis lösen bei mir Errors aus:
Beispielsweise bei Eisvogel:

```
Error producing PDF.
! Argument of \str_uppercase:n has an extra }.
<inserted text> 
                \par 
l.224   \setmainlanguage[]{}

make: *** [Makefile:192] Fehler 43

```

Ich habe das Eisvogel Template als Standalone schon benutzt und es hat geklappt. Jedenfalls optisch war ich mit dem Output recht zufrieden. Nur hat die Vorlage nicht das richtige Format für eine Thesis gehabt (PrüferIn etc). 
Da habe ich deins gefunden und der simple style hat auch direkt auch Anhieb kompiliert. Es ist meine erste Arbeit in der Form, deshalb konnte ich im Makefile einige Anpassungen erkennen und habe angefangen sie zu ändern. 
Dazu zählen neben den og noch abwechselnd zum Original: ` documentetype=report`.
Biserh funktioniert das auch nur die Hervorhebung wäre eben noch ein Träumchen. Eine einfache Hervorhebung wäre schon völig ausreichend.

lg

**Reply by cagix** [reply, marked as answer]
@moerk0 Sorry für die späte Antwort. Bin grad komplett unter Wasser mit Prüfungen ...

Pandoc übersetzt die Markdowndateien nach LaTeX und die Code-Snippets werden wg. der Option `--listings` mit dem `listings`-Paket formatiert. 

Aus 

    Blockcode:

    ```python
    def foo():
      """ Wuppie! """
      pass
    ```

    Und Inline-Code sieht `void foo();` so aus.

wird

```latex
\begin{document}

Blockcode:

\begin{lstlisting}[language=Python]
def foo():
  """ Wuppie! """
  pass
\end{lstlisting}

Und Inline-Code sieht \passthrough{\lstinline!void foo();!} so aus.

\end{document}
```

Man kann dem Inline-Code noch die gewünschte Sprache mitgeben: 

```markdown
  Und Inline-Code sieht `void foo();`{.java} so aus.
```

Was kannst Du von hier aus tun? 

Du könntest die Option `--listings` abschalten - dann nimmt Pandoc eine eigene Formatierung, die sich iirc auch auf Inline-Code erstreckt. Allerdings verlierst Du dabei auch die Möglichkeit, eine Caption o.ä. für Code-Blöcke zu vergeben.

Alternativ könntest Du Dir eigene Definitionen für das `listings`-Paket basteln und einbinden, indem Du diese in `include-header.tex` mit aufnimmst. Hier ein Eintrag auf [Stackexchange](https://tex.stackexchange.com/questions/357227/adding-background-color-to-verb-or-lstinline-command-without-colorbox)[^1] und ein Link auf die Doku von [`listings`](https://ctan.mirror.garr.it/mirrors/ctan/macros/latex/contrib/listings/listings.pdf).

[^1]: Die Diskussion auf SE unterscheidet dabei aber nicht zw. Code-Blöcken und Inline-Code, d.h. bei den dort vorgeschlagenen Lösungen würden auch die Code-Blöcke einen grauen Hintergrund bekommen. Du willst ja vermutlich nur den Inline-Code extra hervorheben, oder?

**Reply by cagix** [reply]
@moerk0 Für eine Analyse des Fehlers bräuchte ich ein _minimum working example_, um das reproduzieren und analysieren zu können.

**Reply by cagix** [reply]
@moerk0 Noch eine Idee, wenn die Konfiguration von `listings` zu nix führt: Ich hatte mal für Foliensätze eine Art Workaround gebastelt, damit nur die Inline-Code-Snippets farblich hervorgehoben wurden.

Im Markdown bleibt es bei der üblichen Schreibweise.

Für Pandoc muss noch ein kleiner Lua-Filter definiert und eingebunden werden:

```lua
-- wrap inline code in `inlinecode` LaTeX command
function Code(el)
    return { pandoc.RawInline("latex", "\\inlinecode{"), el, pandoc.RawInline("latex", "}") }
end
```

Und Du brauchst die passenden LaTeX-Definitionen:

```latex
\usepackage{xcolor}
\definecolor{code}{HTML}{F2F2F2}

\usepackage{tcolorbox}
\newtcbox{\inlinecode}{on line,frame hidden,colback=code,boxsep=0pt,left=1pt,right=1pt,top=2pt,bottom=2pt,boxrule=0pt}
```

Hier wird dann durch den Filter jedes Auftreten eines Inline-Code-Snippets in einen `\inlinecode{}` verpackt, und wenn LaTeX dann läuft, wird mit `tcolorbox` eine farbige Box gebastelt.

**Reply by taste5** [reply]
Moin, kein Problem wenn es länger dauert, ich bin ja schon froh mit dir zu kommunizieren.
Habe listings jetzt raus genommen und es hat sich was verändert, der Hintergrund wurde zwar nicht grau , aber die Typographie hat sich zumindest schonmal geändert. … Für weitere Änderungen an diesem kleinen Detail habe ich leider nicht mehr Zeit.

Deshalb wird es wohl bei einer Typographischen Abwandlung bleiben.  Danke dir.

**Reply by cagix** [reply]
Oh, schade! Aber wenn das schon reicht, dann passt es ja :)  Pandoc versucht eine Art Syntaxhighlighting auch für Inline-Code. Hmmm.

**Reply by cagix** [reply]
Der Weg mit dem Filter sollte eigentlich auch schnell gehen - Du musst den Filter nur in eine Datei `wuppie.lua` packen und im Makefile an der passenden Stelle die Option `-L wuppie.lua` einbauen :) ... und natürlich die Tex-Definitionen in `include-header.tex` aufnehmen. 

Aber ich habe das auch nur aus dem Kopf geschrieben, kann sein, dass es nicht direkt auf Anhieb läuft. Hmmm.

**Comment by taste5** [comment]
> @moerk0 Sorry für die späte Antwort. Bin grad komplett unter Wasser mit Prüfungen ...
> 
> Pandoc übersetzt die Markdowndateien nach LaTeX und die Code-Snippets werden wg. der Option `--listings` mit dem `listings`-Paket formatiert.
> 
> Aus
> 
> ```
> Blockcode:
> 
> ```python
> def foo():
>   """ Wuppie! """
>   pass
> ```
> 
> Und Inline-Code sieht `void foo();` so aus.
> ```
> 
> wird
> 
> ```
> \begin{document}
> 
> Blockcode:
> 
> \begin{lstlisting}[language=Python]
> def foo():
>   """ Wuppie! """
>   pass
> \end{lstlisting}
> 
> Und Inline-Code sieht \passthrough{\lstinline!void foo();!} so aus.
> 
> \end{document}
> ```
> 
> Man kann dem Inline-Code noch die gewünschte Sprache mitgeben:
> 
> ```
>   Und Inline-Code sieht `void foo();`{.java} so aus.
> ```
> 
> Was kannst Du von hier aus tun?
> 
> Du könntest die Option `--listings` abschalten - dann nimmt Pandoc eine eigene Formatierung, die sich iirc auch auf Inline-Code erstreckt. Allerdings verlierst Du dabei auch die Möglichkeit, eine Caption o.ä. für Code-Blöcke zu vergeben.
> 
> Alternativ könntest Du Dir eigene Definitionen für das `listings`-Paket basteln und einbinden, indem Du diese in `include-header.tex` mit aufnimmst. Hier ein Eintrag auf [Stackexchange](https://tex.stackexchange.com/questions/357227/adding-background-color-to-verb-or-lstinline-command-without-colorbox)[1](#user-content-fn-1-85868fab98f138e195ac0fae56b333d6) und ein Link auf die Doku von [`listings`](https://ctan.mirror.garr.it/mirrors/ctan/macros/latex/contrib/listings/listings.pdf).
> ## Footnotes
> 
>     1. Die Diskussion auf SE unterscheidet dabei aber nicht zw. Code-Blöcken und Inline-Code, d.h. bei den dort vorgeschlagenen Lösungen würden auch die Code-Blöcke einen grauen Hintergrund bekommen. Du willst ja vermutlich nur den Inline-Code extra hervorheben, oder? [↩](#user-content-fnref-1-85868fab98f138e195ac0fae56b333d6)

**Comment by taste5** [comment]
Bezüglich des Eisvogel Fehlers folgende Steops to reproduce:

1. Clone the repo
2. try to run `make eisvogel` with default settings
```
Unable to find image 'pandoc-thesis:latest' locally
docker: Error response from daemon: pull access denied for pandoc-thesis, repository does not exist or may require 'docker login': denied: requested access to the resource is denied.

```
3. change `PANDOC` from docker to `pandoc`  (Version 3.1.11.1) in the Makefile
4. run `make eisvogel` again

```
Error producing PDF.
! Package babel Error: Unknown option 'ngerman'. Either you misspelled it
(babel)                or the language definition file ngerman.ldf was not foun
d.

l.4286     \ExecuteOptions{\bbl@opt@main}
make: *** [Makefile:191: thesis.pdf] Fehler 43
```

So as to my findings the error is pointing somewhere to the german language pack.

But apart from that there is another thing I consider strange. After freshly cloning your repository I cannot even build a default document running  `make simple`. I expect it to generate some sort of generic Document. But even the simple build threw an error:

```
! Undefined control sequence.
l.275     \newpairofpagestyles

make: *** [Makefile:191: thesis.pdf] Fehler 43

```

Just informing you here. No need to immedeatly take action. Maybe I oversaw anything or do not understand the concept or my pandoc is outdated. Else it may be something in the repo that needs further examination... If you need more Information hit me up.

And good luck for your exams!


PS I will consider the ommit of the listings option as temporary quick fix answer. Thanks!

**Reply by cagix** [reply]
@moerk0 let's see...

(2)
> Unable to find image 'pandoc-thesis:latest' locally

you just need to build the docker container first :) ... `make docker`.

(3) 
> change PANDOC from docker to pandoc (Version 3.1.11.1) in the Makefile

usually it is a smart idea to use the current pandoc version. unfortunately, pandoc's internal template keeps changing (to a greater or lesser degree), so that i have to wait for eisvogel to adapt each time and then have to adapt my templates here as well. that's why i have defined a specific pandoc version together with a specific version of eisvogel in the docker container. 

however, you're using pandoc directly. since the current pandoc no longer matches the templates (both eisvogel and mine), there are various errors with all kinds of tex packages :/ 

---

i am currently thinking about changing the internal architecture of the system. on this occasion, the current pandoc and eisvogel versions will also be adapted. (see #23, #22, #21, #27, ...)
